hack/dockerfiles/generated-files.Dockerfile

```diff
@@ -0,0 +1,29 @@
+# protoc is dynamically linked to glibc to can't use golang:1.9-alpine
+FROM golang:1.9 AS gobuild-base
+ARG PROTOC_VERSION=3.1.0
+ARG GOGO_VERSION=master
+RUN apt-get update && apt-get install -y \
+ git \
+ unzip \
+ && true
+RUN wget -q https://github.com/google/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-linux-x86_64.zip && unzip protoc-${PROTOC_VERSION}-linux-x86_64.zip -d /usr/local
+RUN protoc --version
+
+RUN go get -d github.com/gogo/protobuf/protoc-gen-gogofaster \
+  && cd /go/src/github.com/gogo/protobuf \
+        && git checkout -q $GOGO_VERSION \
+        && go install ./protoc-gen-gogo ./protoc-gen-gogofaster ./protoc-gen-gogoslick
+
+WORKDIR /go/src/github.com/moby/buildkit
+COPY . .
+RUN go generate ./...
+
+RUN mkdir /generated-files
+RUN find . -name "*.pb.go" ! -path ./vendor/\* | tar -cf - --files-from - | tar -C /generated-files -xf -
+RUN git status --porcelain -- **/*.pb.go 2>/dev/null | tee /dev/stderr >/generated-files.status
```

Comment:
Why not just let it error (like I think it was in a previous version) with an error message after moving it to the validate stage so that it doesn't affect running `update`. To run the validator you would just build and not run a container.

Reply:
WIll do.

Reply:
The resultant console spew is pretty ugly though:
```
Step 15/15 : RUN diffs=$(git status --porcelain -- **/*.pb.go 2>/dev/null);     if [ -n "$diffs" ] ; then         
echo 'The result of "go generate" differs';         echo; 	echo "$diffs";         echo;         echo 
'Please update with "make generate"';         echo; 	exit 1;     fi
 ---> Running in 710250ab521c
The result of "go generate" differs

 M solver/pb/ops.pb.go

Please update with "make generate"

Removing intermediate container 710250ab521c
The command '/bin/sh -c diffs=$(git status --porcelain -- **/*.pb.go 2>/dev/null);     if [ -n "$diffs" ] ; 
then         echo 'The result of "go generate" differs';         echo; 	echo "$diffs";         echo;         
echo 'Please update with "make generate"';         echo; 	exit 1;     fi' returned a non-zero code: 1
Makefile:49: recipe for target 'validate-generated-files' failed
make: *** [validate-generated-files] Error 1
```

The error message is a bit obscured in there compared with having the check in the script:
```
Step 15/15 : RUN diffs=$(git status --porcelain -- **/*.pb.go 2>/dev/null);     if [ -n "$diffs" ] ; then         echo 'The result of "go generate" differs';         echo; 	echo "$diffs";         echo;         echo 'Please update with "make generate"';         echo; 	exit 1;     fi
 ---> Running in b5566d8df0b1
The result of "go generate" differs

 M solver/pb/ops.pb.go

Please update with "make generate"

Removing intermediate container b5566d8df0b1
The command '/bin/sh -c diffs=$(git status --porcelain -- **/*.pb.go 2>/dev/null);     if [ -n "$diffs" ] ; then         echo 'The result of "go generate" differs';         echo; 	echo "$diffs";         echo;         echo 'Please update with "make generate"';         echo; 	exit 1;     fi' returned a non-zero code: 1
Makefile:49: recipe for target 'validate-generated-files' failed
make: *** [validate-generated-files] Error 1

```

I'd be inclined to leave it in the script.

Reply:
What's the difference, other than that one is wrapped? If we want to make it more readable I'd suggest adding this command to a script in the previous command and then just running `RUN validate-generated` as the last step (or in the container if you think that is better).

Reply:
Oops, looks like I cut-and-pasted the wrong "good" case. The current case is actually:

```
Successfully built 2b15397be9d0
++ cat /tmp/docker-iidfile.rfrZoSDmlD
+ iid=sha256:2b15397be9d03644a2bab51e84a64ed025422e2c052d6cd3a24f2c0331480b1c
++ docker run --rm sha256:2b15397be9d03644a2bab51e84a64ed025422e2c052d6cd3a24f2c0331480b1c cat /generated-files.status
+ diffs=' M solver/pb/ops.pb.go'
+ '[' ' M solver/pb/ops.pb.go' ']'
+ set +x
The result of "go generate" differs

 M solver/pb/ops.pb.go

Please update with "make generate"

Makefile:49: recipe for target 'validate-generated-files' failed
make: *** [validate-generated-files] Error 1
```

Which in particular lacks the escaped/wrapped shell code which appears after the error message in the other case. In fact with it in the dockerfile the error message is bookended by two copies of that same escaped/wrapped shell code which hides it somewhat.

I'll see how putting it in a script looks.

Reply:
Looks good (the error message is red too):

```
Step 15/15 : RUN ./hack/validate-generated-files check
 ---> Running in fdcbdfae3244
The result of "go generate" differs

 M solver/pb/ops.pb.go

Please update with "make generate"

Removing intermediate container fdcbdfae3244
The command '/bin/sh -c ./hack/validate-generated-files check' returned a non-zero code: 1
Makefile:49: recipe for target 'validate-generated-files' failed
make: *** [validate-generated-files] Error 1
```